### PR TITLE
Improve MultiSet's Foldable instances

### PR DESCRIPTION
**Pull request by sjakobi**
TODO:

* [x] Complete MultiSet Foldable implementation
* [x] Add tests (with ~~[`quickcheck-classes`](https://hackage.haskell.org/package/quickcheck-classes)?~~ `checkers`!)
* [x] ~~Do the same for the IntMultiSet's Foldable implementation~~ IntMultiSet of course doesn't have a Foldable instance!
* [x] Fix the doctests

### CODE REVIEW

File: Data/MultiSet.hs [hunk 1]
```diff
@@ -241,7 +258,7 @@ null = Map.null . unMS
 
 -- | /O(n)/. The number of elements in the multiset.
```
Comment (sjakobi) [inline Data/MultiSet.hs]: This should be `O(# distinct elements)` I guess.

File: multiset.cabal [hunk 2]
```diff
@@ -25,7 +25,7 @@ Library
   include-dirs:       include
   default-extensions: CPP
   ghc-options:        -Wall
-  build-depends:      containers >= 0.5, base >= 4 && < 5, deepseq >=1.2 && <1.5
+  build-depends:      containers >= 0.5.4, base >= 4 && < 5, deepseq >=1.2 && <1.5
```
Comment (sjakobi) [inline multiset.cabal]: Does that new bound seem acceptable? v0.5.4 was released in 2014.

Comment (twanvl) [reply multiset.cabal]: The bound on base probably also needs to be updated, since your changes seem to require Semigroup => Monoid.

**Comment by sjakobi** [comment]
Regarding the tests, it turns out that `MultiSet` can't have an `Arbitrary1` instance as required by `quickcheck-classes`. The problem is the `Ord` constraint on the elements.

I'll try `checkers` next, although `checkers` doesn't offer [`Foldable` laws](https://github.com/conal/checkers/issues/22) so far.

File: test/multiset-properties.hs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  arbitrary = fromList <$> arbitrary
+
+instance Eq a => EqProp (MultiSet a) where
+  (=-=) = Test.QuickCheck.Checkers.eq
+
+instance Eq a => EqProp (Sum a) where
+  (=-=) = Test.QuickCheck.Checkers.eq
+
+instance EqProp Integer where
+  (=-=) = Test.QuickCheck.Checkers.eq
```
Comment (sjakobi) [inline test/multiset-properties.hs]: TODO: Remove this instance once https://github.com/conal/checkers/pull/28 has been merged. Also fix the lower bound on `checkers` then.

File: multiset.cabal (hunk 2 above)
Comment (sjakobi) [reply multiset.cabal]: > Alternatively I can add a bit of CPP.

I ended up doing that. This patch builds with GHC-7.8 now.

**Comment by sjakobi** [comment]
@twanvl Could you give this PR another review?

Note that the existing `-fno-warn-amp` options cause some compatibility issues. #40 Takes care of those.

**Comment by sjakobi** [comment]
@twanvl This PR [passes](https://www.travis-ci.org/twanvl/multiset/builds/525385160) CI now. Can you give it another review?

**Comment by twanvl** [comment]
Why are all the changes to travis-ci needed to improve the Foldable instance?

**Comment by twanvl** [comment]
Also, perhaps the definition of `foldr` should now be changed to `Foldable.foldr`. Or perhaps it makes sense to move the current definition of foldr into `Foldable.foldr` (as is done in this patch for `foldl`)

**Comment by sjakobi** [comment]
> Why are all the changes to travis-ci needed to improve the Foldable instance?

That was a bit of a rabbit hole. Mainly the new testsuite depends on `checkers` which isn't compatible with GHC <= 7.8.

File: Data/MultiSet.hs [hunk 4]
```diff
@@ -207,8 +207,25 @@ instance Ord a => Semigroup (MultiSet a) where
...
+    foldr = foldr
+    -- foldr'
+    -- foldl
+    -- foldl'
+    -- foldr1
+    -- foldl1
+    toList = Map.foldMapWithKey (\x n -> replicate n x) . unMS
+    null = null
+    length = size
+    elem x = elem x . distinctElems
```
[the remark this answers is not in the document]
Comment (sjakobi) [reply Data/MultiSet.hs]: `member` has an `Ord`-constraint that `elem` cannot have.

File: Data/MultiSet.hs [hunk 5]
```diff
@@ -207,8 +211,42 @@ instance Ord a => Semigroup (MultiSet a) where
       | otherwise = MS $ Map.map (* fromIntegral n) m
 #endif
 
+-- | Note that 'elem' is slower than 'member'.
 instance Foldable.Foldable MultiSet where
-    foldr = fold
+    foldr = foldr
```
Comment (sjakobi) [inline Data/MultiSet.hs]: > Change Foldable.foldr to use top level foldr function.

The `foldr` on the right already means the top-level `foldr` from this module.

```suggestion
    foldr = fold
```

Would this be less confusing?

File: Data/MultiSet.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sjakobi) [reply Data/MultiSet.hs]: Actually we could add an `Ord` constraint to the entire instance. How about that?

[the remark this answers is not in the document]
Comment (sjakobi) [reply Data/MultiSet.hs]: > Actually we could add an `Ord` constraint to the entire instance.

Actually we can can't – the key type doesn't appear in the instance head so we can't put a constraint on it!

**Comment by sjakobi** [comment]
@twanvl I've tried to address your comments as best as I could.

IMHO this is good to go now.

**Comment by sjakobi** [comment]
Bump! :)

**Comment by sjakobi** [comment]
Bump! :)

**Comment by sjakobi** [comment]
Ping @twanvl! :)

**Pull request merged**